src/main/java/com/microsoft/sqlserver/jdbc/SharedTimer.java

```diff
@@ -111,20 +111,16 @@ public void removeRef() {
      * When the caller is finished with the SharedTimer it must be released via {@link#removeRef}
      */
     public static SharedTimer getTimer() {
-        SharedTimer result = instance;
```

Comment:
Thank you for the changes. For now, we'd like to resolve the race condition, but keep everything else the same, so keeping the extra SharedTimer variable etc. So, the final PR will have getAndIncrement() moved as well as the extra test. If you're able to do that for us, we can move forward with merging this.

Reply:
You can't just move the `getAndIncrement()`.
I can keep the local var, but it makes no sense:
```
    public static SharedTimer getTimer() {
        LOCK.lock();
        try {
            SharedTimer result = instance;
            if (result == null) {
                instance = result = new SharedTimer();
            }
            result.refCount.getAndIncrement();
            return result;
        } finally {
            LOCK.unlock();
        }
    }
```

Like this? If you insist, fine.

Reply:
Maybe I'm misunderstanding. The race condition comes about because the reference count has not been placed correctly in regard to the locking, so the only thing that needs to be fixed here is its placement. We still want the extra null check before the locking, and we still want to abstract away from instance and compare everything to result as instance is volatile. Below is what I'm looking for, just the refCount logic being moved. I would help you out here and push directly to the PR, but I'm unable to do so, pushing to user remotes doesn't always work.

```public static SharedTimer getTimer() {
        SharedTimer result = instance;
        if (result == null) {
            LOCK.lock();
            try {
                result = instance;
                if (result == null) {
                    instance = result = new SharedTimer();
                }
                result.refCount.getAndIncrement();
            } finally {
                LOCK.unlock();
            }
        }
        return result;
    }
```

Reply:
Your code doesn't increment the counter when the instance exists.

Reply:
You're right, I copied over the wrong thing, I meant to increase the refCount for instance.

```public static SharedTimer getTimer() {
        SharedTimer result = instance;
        if (result == null) {
            LOCK.lock();
            try {
                result = instance;
                if (result == null) {
                    instance = result = new SharedTimer();
                }
                instance.refCount.getAndIncrement();
            } finally {
                LOCK.unlock();
            }
        }
        return result;
    }
```

Reply:
Still the same...